source/Cosmos.HAL2/BlockDevice/AHCI.cs

```diff
@@ -0,0 +1,380 @@
+                {
+                    PortRegisters xPortReg = new PortRegisters((uint)mABAR + 0x100, (uint)xPort);
+                    PortType PortType = CheckPortType(xPortReg);
+                    xPortReg.mPortType = PortType;
+                    var xPortString = "0:" + ((xPort.ToString().Length <= 1) ? xPort.ToString().PadLeft(1, '0') : xPort.ToString());
+                    if (PortType == PortType.SATA) // If Port type was SATA.
+                    {
+                        mAHCIDebugger.Send("Initializing Port " + xPortString + " with type SATA");
+                        PortRebase(xPortReg, (uint)xPort);
+                        var xSATAPort = new SATA(xPortReg);
+                        mPorts.Add(xSATAPort);
+                    }
+                    else if (PortType == PortType.SATAPI) // If Port type was SATAPI.
+                    {
+                        mAHCIDebugger.Send("Initializing Port " + xPortString + " with type Serial ATAPI");
+                        //PortRebase(xPortReg, (uint)xPort);
+                        //var xSATAPIPort = new SATAPI(xPortReg);
+                        //mPorts.Add(xSATAPIPort);
+                    }
+                    else if (PortType == PortType.SEMB) // If Port type was SEMB.
+                    {
+                        mAHCIDebugger.Send("SEMB Drive at port " + xPortString + " found, which is not supported yet!");
+                    }
+                    else if (PortType == PortType.PM) // If Port type was Port Mulitplier.
+                    {
+                        mAHCIDebugger.Send("Port Multiplier Drive at port " + xPortString + " found, which is not supported yet!");
+                    }
+                    else if (PortType != PortType.Nothing)
+                        mAHCIDebugger.Send("Unknown drive found with signature: 0x" + xPortReg.SIG);// If Implemented Port value was not zero and non of the above.
+                }
+                xImplementedPort >>= 1;
+            }
+        }
+
+        private PortType CheckPortType(PortRegisters aPort)
+        {
+            var xIPM = (InterfacePowerManagementStatus)((aPort.SSTS >> 8) & 0x0F);
+            var xSPD = (CurrentInterfaceSpeedStatus)((aPort.SSTS >> 4) & 0x0F);
+            var xDET = (DeviceDetectionStatus)(aPort.SSTS & 0x0F);
+            var xSignature = aPort.SIG;
+
+            // Check if not reading wrong data!
+            if (xIPM != InterfacePowerManagementStatus.Active)
+                return PortType.Nothing;
+            if (xDET != DeviceDetectionStatus.DeviceDetectedWithPhy)
+                return PortType.Nothing;
+
+            switch (xSignature >> 16)
+            {
+                case 0xEB14: return PortType.SATAPI;
+                case 0xC33C: return PortType.SEMB;
+                case 0x9669: return PortType.PM;
+                case 0xFFFF: return PortType.Nothing;
+                default: return PortType.SATA;
```

Comment:
Use enums here too for 0xEB14, 0xC33C and so on... make PortType.SATA explicit and put an exception on the default as - at that point - you are again in an "impossible" situation.

Reply:
well, if the code didn't fall in the `if` trap, then of course this port is a sata port

Reply:
The port is active and detected